src/Spectre.libClassifier.Tests/OpenCvDatasetTest.cpp

```diff
@@ -0,0 +1,225 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+
+#include <gtest/gtest.h>
+#include <span.h>
+#include <opencv2/core/mat.hpp>
+#include "Spectre.libClassifier/OpenCvDataset.h"
+#include "Spectre.libException/InconsistentArgumentSizesException.h"
+#include "Spectre.libException/OutOfRangeException.h"
+
+namespace
+{
+using namespace Spectre::libClassifier;
+using namespace Spectre::libException;
+
+class OpenCvDatasetInitializationTest : public ::testing::Test
+{
+protected:
+	const std::vector<DataType> data_long{ 0.5f, 0.4f, 0.6f, 1.1f, 1.6f, 0.7f, 2.1f, 1.0f, 0.6f };
+	const std::vector<DataType> data_short{ 0.5f, 0.4f, 0.6f };
+	const std::vector<Label> labels{ 3, 7, 14 };
+	const std::vector<Label> labels_too_long{ 3, 7, 14, 5 };
+};
+
+TEST_F(OpenCvDatasetInitializationTest, correct_dataset_opencv_initialization_col_size_one)
+{
+	EXPECT_NO_THROW(OpenCvDataset(data_short, labels));
+}
+
+TEST_F(OpenCvDatasetInitializationTest, correct_dataset_opencv_initialization)
+{
+	EXPECT_NO_THROW(OpenCvDataset(data_long, labels));
+}
+
+TEST_F(OpenCvDatasetInitializationTest, throws_for_inconsistent_size)
+{
+	EXPECT_THROW(OpenCvDataset(data_short, labels_too_long), InconsistentArgumentSizesException);
+}
+
+TEST_F(OpenCvDatasetInitializationTest, correct_dataset_opencv_initialization_from_mat_col_size_one)
+{
+	std::vector<DataType> tmp_data(data_short);
+	std::vector<Label> tmp_labels(labels);
+	cv::Mat mat_data(3, 1, CV_TYPE, tmp_data.data());
+	cv::Mat mat_labels(3, 1, CV_LABEL_TYPE, tmp_labels.data());
+	EXPECT_NO_THROW(OpenCvDataset(mat_data, mat_labels));
+}
+
+TEST_F(OpenCvDatasetInitializationTest, correct_dataset_opencv_initialization_from_mat)
+{
+	std::vector<DataType> tmp_data(data_long);
+	std::vector<Label> tmp_labels(labels);
+	cv::Mat mat_data(3, 3, CV_TYPE, tmp_data.data());
+	cv::Mat mat_labels(3, 1, CV_LABEL_TYPE, tmp_labels.data());
+	EXPECT_NO_THROW(OpenCvDataset(mat_data, mat_labels));
+}
+
+TEST_F(OpenCvDatasetInitializationTest, throws_for_inconsistent_size_from_mat)
+{
+	std::vector<DataType> tmp_data(data_long);
+	std::vector<Label> tmp_labels(labels_too_long);
+	cv::Mat mat_data(3, 3, CV_TYPE, tmp_data.data());
+	cv::Mat mat_labels(4, 1, CV_LABEL_TYPE, tmp_labels.data());
+	EXPECT_THROW(OpenCvDataset(mat_data, mat_labels), InconsistentArgumentSizesException);
+}
+
+class OpenCvDatasetTest : public ::testing::Test
+{
+public:
+	OpenCvDatasetTest()
+	{
+
+	}
+
+    static bool compareCVMats(cv::Mat mat1, cv::Mat mat2)
```

Comment:
`cv::Mat` is just a view over data, it can be copied at no expense (it does not own the data). It's lightweight (just like pointer + some sizes).